include/transport.h

```diff
@@ -0,0 +1,82 @@
+/* Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MIXERCLIENT_TRANSPORT_H
+#define MIXERCLIENT_TRANSPORT_H
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "google/protobuf/stubs/status.h"
+#include "mixer/api/v1/service.pb.h"
+
+namespace istio {
+namespace mixer_client {
+
+// A stream write interface implemented by transport layer
+// It will be called by Mixer client.
+template <class RequestType>
+class WriteInterface {
+ public:
+  // Write a request message.
+  virtual void Write(const RequestType&) = 0;
```

Comment:
Such error should cause stream to be closed.  So reader->OnClose(status) will be called; a single point of close/error handling.